### PR TITLE
 Add update secret route

### PR DESCRIPTION
**Pull request by minz1027**
- Add `update` route for secrets
- Only user can `update` a secret
- Modify `canAccess` helper function to take in required permission
- Update README

Issue #232

### CODE REVIEW

**Comment by d2lam** [comment]
Blocked by https://github.com/screwdriver-cd/data-schema/pull/62

File: plugins/secrets/update.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            const credentials = request.auth.credentials;
+            const canAccess = request.server.plugins.secrets.canAccess;
+
+            return factory.get(request.params.id)
+                .then(secret => {
+                    if (!secret) {
+                        throw boom.notFound('Secret does not exist');
+                    }
+
+                    Object.keys(request.payload).forEach(key => {
```
Comment (d2lam) [inline plugins/secrets/update.js]: Should probably move this to after canAccess, right before `secret.update()`. No point of doing this if user doesn't have permission.

Comment (minz1027) [inline plugins/secrets/update.js]: fixed

File: plugins/secrets/update.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            const credentials = request.auth.credentials;
+            const canAccess = request.server.plugins.secrets.canAccess;
+
+            return factory.get(request.params.id)
+                .then(secret => {
+                    if (!secret) {
+                        throw boom.notFound('Secret does not exist');
+                    }
+
+                    // Make sure that user/build has permission before updating
```
Comment (stjohnjohnson) [inline plugins/secrets/update.js]: builds cannot update secrets

File: plugins/secrets/update.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                    return canAccess(credentials, secret)
+                        .then(() => {
```
Comment (stjohnjohnson) [inline plugins/secrets/update.js]: You're not doing anything with the result.  But even so, that check only checks for write access (which is not the same as admin for create).  I would assume you would need admin to update if you need admin to create.

File: plugins/secrets/update.js (hunk 2 above)
Comment (minz1027) [inline plugins/secrets/update.js]: fixed

File: plugins/secrets/update.js (hunk 3 above)
Comment (stjohnjohnson) [inline plugins/secrets/update.js]: What do you think should be the appropriate permission level?  Who should be creating/updating/deleting secrets?  Whichever we pick, we should document.

Comment (d2lam) [inline plugins/secrets/update.js]: Hmmm I think admin access for create, update and remove. write access for get. We can modify canAccess to do this. 
If we want to simplify we can make everything requires admin access.

Comment (minz1027) [inline plugins/secrets/update.js]: @stjohnjohnson I agree with admin access for `create`, `update` and `remove`. write access for `get` and `list`. Have modified the `canAccess` function to take care of these 2 permission levels. Plz help review again. :)

File: plugins/secrets/index.js [hunk 4]
```diff
@@ -42,6 +49,11 @@ exports.register = (server, options, next) => {
                     }
 
                     return user.getPermissions(pipeline.scmUrl).then(permissions => {
+                        if (permission === 'admin' && !permissions.admin) {
```
Comment (d2lam) [inline plugins/secrets/index.js]: You can do this so you don't need to split into multiple checks for admin and push:
```js 
if (!permisions[requiredPerm]) {
   throw boom.forbidden(`User ${username} does not have ${requiredPerm} access to this repo`)
``` 
where `requiredPerm` is the param passed in.

Comment (minz1027) [inline plugins/secrets/index.js]: fixed

File: plugins/secrets/README.md [hunk 5]
```diff
@@ -26,7 +26,11 @@ server.register({
 ```
 
 ### Routes
-#### Get a single secret
+`Get` requires **write** permission.
```
Comment (d2lam) [inline plugins/secrets/README.md]: permission to the repository to be more clear?

File: plugins/secrets/README.md [hunk 6]
```diff
@@ -26,7 +26,11 @@ server.register({
...
+
+`Create`, `Remove` and `Update` require **admin** permission.
```
Comment (d2lam) [inline plugins/secrets/README.md]: same here

File: plugins/secrets/README.md (hunk 5 above)
Comment (minz1027) [inline plugins/secrets/README.md]: fixed

**Pull request merged**